jcl/src/java.base/share/classes/java/lang/Class.java

```diff
@@ -1767,19 +1743,11 @@ Method getMethodHelper(
 	if (methodList == null) {
 		// getDeclaredPublicMethods() has to go through all methods anyway
 		Method cachedMethod = lookupCachedMethod(name, parameterTypes);
-		if (cachedMethod != null) {
-			if (searchAllMethods && forDeclaredMethod && (cachedMethod.getDeclaringClass() == this)) {
-				return cachedMethod;
-			} else {
-				boolean isPublic = Modifier.isPublic(cachedMethod.getModifiers());
-				if (searchAllMethods) {
-					if (!forDeclaredMethod && isPublic) {
-						return cachedMethod;
-					}
-				} else if (searchPublicMethodOnly == isPublic) {
-					return cachedMethod;
-				}
-			}
+		if ((cachedMethod != null)
+			&& ((forDeclaredMethod && (cachedMethod.getDeclaringClass() == this))
+			|| (!forDeclaredMethod && (Modifier.isPublic(cachedMethod.getModifiers()) || !publicOnly)))
+		) {
+			return cachedMethod;
 		}
```

Comment:
I find this rather difficult to follow, and we need test `forDeclaredMethod` at most once; also we don't need to call `isPublic()` unless `!publicOnly`:
```java
		if (cachedMethod != null) {
			if (forDeclaredMethod) {
				if (cachedMethod.getDeclaringClass() == this) {
					return cachedMethod;
				}
			} else {
				if (!publicOnly || Modifier.isPublic(cachedMethod.getModifiers())) {
					return cachedMethod;
				}
			}
		}
```

Reply:
Updated.